librad/src/git/storage.rs

```diff
@@ -522,94 +619,206 @@ impl Storage<WithSigner> {
-            let refspecs =
-                Refspec::rad_refs(namespace.clone(), from, tracked_trans.iter().cloned())
-                    .map(|spec| spec.to_string())
-                    .collect::<Vec<String>>();
-            tracing::debug!(refspecs = ?refspecs, "Fetching rad/refs");
-            remote.fetch(&refspecs, Some(&mut self.fetch_options()), None)?;
-        }
+        let rad_refs = self.rad_refs(&urn)?;
+        let transitively_tracked = rad_refs.remotes.flatten().collect::<HashSet<&PeerId>>();
 
-        // Read the signed refs of all known remotes, and compare their `heads`
-        // against the advertised refs. If signed and advertised branch head
-        // match, non-fast-forwards are permitted. Otherwise, the branch is
-        // skipped.
+        fetcher.fetch(
+            transitively_tracked,
+            |peer| self.rad_refs_of(&urn, peer),
+            |peer| self.certifiers_of(&urn, peer),
+        )?;
+
+        // Symref any certifiers from `remote_peer`, ie. for all valid refs in
+        // the remotes's `rad/ids/*`, create a symref in the _local_ `rad/ids/*`
+        // pointing to the `rad/id` in the respective top-level namespace.
         {
-            let remote_heads: HashMap<&str, git2::Oid> = remote
-                .list()?
-                .iter()
-                .map(|rhead| (rhead.name(), rhead.oid()))
-                .collect();
-
-            let refspecs = Refspec::fetch_heads(
-                namespace.clone(),
-                remote_heads,
-                tracked_trans.iter().cloned(),
-                from,
-                |peer| self.rad_refs_of(urn, peer),
-                |peer| self.certifiers_of(urn, peer),
+            References::from_globs(
+                &self.backend,
+                &[format!(
+                    "refs/namespaces/{}/refs/remotes/{}/rad/ids/*",
+                    &urn.id, &remote_peer
+                )],
             )?
-            .map(|spec| spec.to_string())
-            .collect::<Vec<String>>();
-
-            tracing::debug!(refspecs = ?refspecs, "Fetching refs/heads");
-            remote.fetch(&refspecs, Some(&mut self.fetch_options()), None)?;
+            .names()
+            .try_for_each(|certifier_ref| {
+                let certifier_ref = certifier_ref?;
+                match certifier_ref.parse::<Hash>() {
+                    Err(_) => Ok(()),
+                    Ok(certifier_hash) => {
+                        let certifier_here = Reference::rad_certifier(
+                            urn.id.clone(),
+                            &RadUrn::new(
+                                certifier_hash.clone(),
+                                uri::Protocol::Git,
+                                uri::Path::empty(),
+                            ),
+                        );
+                        let certifier_id = Reference::rad_id(certifier_hash);
+                        self.backend
+                            .reference_symbolic(
```

Comment:
I'm quite fond of these typed `Reference`s. I wonder if we could extend that API a bit with `SymReferences` which could take two `Reference`s? 

It might also be useful to have a method on them, say `create(self, repo: git2::Repostiory) -> Result<(), Error>` that handles the `to_string()` calling.

What do you think?

Reply:
Ya, patches welcome :) It's probably that one would want a constructor with named fields, as the options are not only boolean-blind, but also what's pointing to what blind.